File: src/services/trustService.ts
import { apiClient } from './api';
import { TrustTransaction } from '../types';
import { societyStore } from './store';

export interface TrustSummary {
  userId: number | string;
  userName: string;
  trustPoints: number;
  trustLevel: string;
  nextTrustLevel?: string;
  pointsToNextLevel?: number;
  nextLevelThreshold?: number;
  progressPercentage?: number;
  itemsShared?: number;
  successfulBorrows?: number;
  eventsOrganized?: number;
  verifiedSafetyReports?: number;
}

export const calculateTrustLevel = (points: number) => {
  if (points >= 200) {
    return {
      levelTitle: 'Community Leader',
      nextLevelTitle: undefined,
      pointsToNextLevel: 0,
      nextLevelThreshold: 200,
      progressPercentage: 100,
    };
  } else if (points >= 100) {
    const threshold = 200;
    const needed = Math.max(0, threshold - points);
    const progress = Math.min(100, Math.round((points / threshold) * 100));
    return {
      levelTitle: 'Community Champion',
      nextLevelTitle: 'Community Leader',
      pointsToNextLevel: needed,
      nextLevelThreshold: threshold,
      progressPercentage: progress,
    };
  } else if (points >= 50) {
    const threshold = 100;
    const needed = Math.max(0, threshold - points);
    const progress = Math.min(100, Math.round((points / threshold) * 100));
    return {
      levelTitle: 'Trusted Neighbor',
      nextLevelTitle: 'Community Champion',
      pointsToNextLevel: needed,
      nextLevelThreshold: threshold,
      progressPercentage: progress,
    };
  } else if (points >= 25) {
    const threshold = 50;
    const needed = Math.max(0, threshold - points);
    const progress = Math.min(100, Math.round((points / threshold) * 100));
    return {
      levelTitle: 'Active Neighbor',
      nextLevelTitle: 'Trusted Neighbor',
      pointsToNextLevel: needed,
      nextLevelThreshold: threshold,
      progressPercentage: progress,
    };
  } else {
    const threshold = 25;
    const needed = Math.max(0, threshold - points);
    const progress = Math.min(100, Math.round((Math.max(0, points) / threshold) * 100));
    return {
      levelTitle: 'New Neighbor',
      nextLevelTitle: 'Active Neighbor',
      pointsToNextLevel: needed,
      nextLevelThreshold: threshold,
      progressPercentage: progress,
    };
  }
};

export const trustService = {
  /**
   * GET /api/trust/me
   */
  async getMyTrustSummary(): Promise<TrustSummary> {
    try {
      const response = await apiClient.get<any>('/trust/me');
      const data = response.data?.data || response.data;
      const pts = data.trustPoints ?? 88;
      const computed = calculateTrustLevel(pts);

      return {
        userId: data.userId,
        userName: data.userName,
        trustPoints: pts,
        trustLevel: data.trustLevel || computed.levelTitle,
        nextTrustLevel: data.nextTrustLevel || computed.nextLevelTitle,
        pointsToNextLevel: data.pointsToNextLevel ?? computed.pointsToNextLevel,
        nextLevelThreshold: data.nextLevelThreshold ?? computed.nextLevelThreshold,
        progressPercentage: data.progressPercentage ?? computed.progressPercentage,
        itemsShared: data.itemsShared,
        successfulBorrows: data.successfulBorrows,
        eventsOrganized: data.eventsOrganized,
        verifiedSafetyReports: data.verifiedSafetyReports,
      };
    } catch {
      const pts = 88;
      const computed = calculateTrustLevel(pts);
      return {
        userId: 'usr-1',
        userName: 'Priya Sharma',
        trustPoints: pts,
        trustLevel: computed.levelTitle,
        nextTrustLevel: computed.nextLevelTitle,
        pointsToNextLevel: computed.pointsToNextLevel,
        nextLevelThreshold: computed.nextLevelThreshold,
        progressPercentage: computed.progressPercentage,
        itemsShared: 4,
        successfulBorrows: 7,
        eventsOrganized: 2,
        verifiedSafetyReports: 3,
      };
    }
  },

  /**
   * GET /api/trust/me/history
   */
  async getTrustHistory(userId?: string): Promise<TrustTransaction[]> {
    try {
      const endpoint = userId && userId !== 'me' ? `/trust/user/${userId}/history` : '/trust/me/history';
      const response = await apiClient.get<any>(endpoint);
      const rawList: any[] = response.data?.data || response.data || [];

      return rawList.map((tx: any) => ({
        id: String(tx.id),
        userId: String(tx.userId || userId || 'me'),
        userName: tx.userName,
        apartmentNumber: tx.apartmentNumber,
        points: tx.points,
        reason: tx.reason,
        referenceType: tx.referenceType,
        referenceId: tx.referenceId,
        category: 'RESOURCE_SHARING',
        timestamp: tx.createdAt || new Date().toISOString(),
      }));
    } catch {
      return societyStore.getTrust();
    }
  },

  /**
   * Admin: GET /api/trust/admin/transactions
   */
  async getAllTransactionsForAdmin(): Promise<TrustTransaction[]> {
    try {
      const response = await apiClient.get<any>('/trust/admin/transactions');
      const rawList: any[] = response.data?.data || response.data || [];

      return rawList.map((tx: any) => ({
        id: String(tx.id),
        userId: String(tx.userId),
        userName: tx.userName || 'Resident',
        apartmentNumber: tx.apartmentNumber || 'Apt',
        points: tx.points,
        reason: tx.reason,
        referenceType: tx.referenceType,
        referenceId: tx.referenceId,
        timestamp: tx.createdAt || new Date().toISOString(),
      }));
    } catch {
      return societyStore.getTrust();
    }
  }
};

export default trustService;

File: backend/src/main/java/com/smartneighbour/service/TrustService.java
package com.smartneighbour.service;

import com.smartneighbour.dto.TrustDto.TrustSummaryResponse;
import com.smartneighbour.dto.TrustDto.TrustTransactionResponse;
import com.smartneighbour.entity.TrustTransaction;
import com.smartneighbour.entity.User;
import com.smartneighbour.exception.AppException;
import com.smartneighbour.repository.TrustTransactionRepository;
import com.smartneighbour.repository.UserRepository;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class TrustService {

    private final TrustTransactionRepository trustTransactionRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;

    @Data
    @AllArgsConstructor
    public static class LevelInfo {
        private String levelTitle;
        private String nextLevelTitle;
        private int pointsToNextLevel;
        private int nextLevelThreshold;
        private int progressPercentage;
    }

    public static LevelInfo calculateLevelInfo(int points) {
        if (points >= 200) {
            return new LevelInfo("Community Leader", null, 0, 200, 100);
        } else if (points >= 100) {
            int threshold = 200;
            int needed = threshold - points;
            int progress = (int) Math.min(100, Math.round((double) points / threshold * 100));
            return new LevelInfo("Community Champion", "Community Leader", needed, threshold, progress);
        } else if (points >= 50) {
            int threshold = 100;
            int needed = threshold - points;
            int progress = (int) Math.min(100, Math.round((double) points / threshold * 100));
            return new LevelInfo("Trusted Neighbor", "Community Champion", needed, threshold, progress);
        } else if (points >= 25) {
            int threshold = 50;
            int needed = threshold - points;
            int progress = (int) Math.min(100, Math.round((double) points / threshold * 100));
            return new LevelInfo("Active Neighbor", "Trusted Neighbor", needed, threshold, progress);
        } else {
            int threshold = 25;
            int needed = Math.max(0, threshold - points);
            int progress = (int) Math.min(100, Math.round((double) Math.max(0, points) / threshold * 100));
            return new LevelInfo("New Neighbor", "Active Neighbor", needed, threshold, progress);
        }
    }

    @Transactional
    public void addPoints(Long userId, int points, String reason, String referenceType, Long referenceId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        int newScore = Math.max(0, user.getTrustPoints() + points);
        user.setTrustPoints(newScore);
        userRepository.save(user);

        TrustTransaction transaction = TrustTransaction.builder()
                .user(user)
                .points(points)
                .reason(reason)
                .referenceType(referenceType)
                .referenceId(referenceId)
                .build();

        trustTransactionRepository.save(transaction);

        // Generate readable notification for trust points
        if (points > 0 && notificationService != null) {
            notificationService.createNotification(
                    userId,
                    "Trust Points Earned 🎉",
                    "You earned +" + points + " Trust Points for: " + reason,
                    "TRUST_AWARD",
                    "/trust-points"
            );
        }
    }

    @Transactional(readOnly = true)
    public TrustSummaryResponse getTrustSummary(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new AppException("User not found", HttpStatus.NOT_FOUND));

        int points = user.getTrustPoints();
        LevelInfo info = calculateLevelInfo(points);

        return TrustSummaryResponse.builder()
                .userId(user.getId())
                .userName(user.getFullName())
                .trustPoints(points)
                .trustLevel(info.getLevelTitle())
                .nextTrustLevel(info.getNextLevelTitle())
                .pointsToNextLevel(info.getPointsToNextLevel())
                .nextLevelThreshold(info.getNextLevelThreshold())
                .progressPercentage(info.getProgressPercentage())
                .itemsShared(user.getItems().size())
                .successfulBorrows(user.getBorrowRequests().size())
                .eventsOrganized(user.getEvents().size())
                .verifiedSafetyReports(user.getSafetyReports().size())
                .build();
    }

    @Transactional(readOnly = true)
    public List<TrustTransactionResponse> getTrustHistory(Long userId) {
        return trustTransactionRepository.findByUserIdOrderByCreatedAtDesc(userId).stream()
                .map(tx -> TrustTransactionResponse.builder()
                        .id(tx.getId())
                        .userId(tx.getUser().getId())
                        .userName(tx.getUser().getFullName())
                        .apartmentNumber(tx.getUser().getApartmentNumber())
                        .points(tx.getPoints())
                        .reason(tx.getReason())
                        .referenceType(tx.getReferenceType())
                        .referenceId(tx.getReferenceId())
                        .createdAt(tx.getCreatedAt())
                        .build())
                .toList();
    }

    @Transactional(readOnly = true)
    public List<TrustTransactionResponse> getAllTrustTransactionsForAdmin() {
        return trustTransactionRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(tx -> TrustTransactionResponse.builder()
                        .id(tx.getId())
                        .userId(tx.getUser().getId())
                        .userName(tx.getUser().getFullName())
                        .apartmentNumber(tx.getUser().getApartmentNumber())
                        .points(tx.getPoints())
                        .reason(tx.getReason())
                        .referenceType(tx.getReferenceType())
                        .referenceId(tx.getReferenceId())
                        .createdAt(tx.getCreatedAt())
                        .build())
                .toList();
    }
}

